server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -265,17 +269,71 @@ protected void doStop() {}
 
     @Override
     protected void doClose() {
-        try {
-            blobStore().close();
-        } catch (Exception t) {
-            logger.warn("cannot close blob store", t);
+        BlobStore store = blobStore;
+        if (store != null) {
+            try {
+                store.close();
+            } catch (Exception t) {
+                logger.warn("cannot close blob store", t);
+            }
+        }
+    }
+
+    // package private, only use for testing
+    BlobContainer getBlobContainer() {
+        return blobContainer;
+    }
+
+    // for test purposes only
+    protected BlobStore getBlobStore() {
+        return blobStore;
+    }
+
+    /**
+     * maintains single lazy instance of {@link BlobContainer}
+     */
+    protected BlobContainer blobContainer() {
+        BlobContainer blobContainer = this.blobContainer;
+        if (blobContainer == null) {
+           synchronized (lock) {
+               blobContainer = this.blobContainer;
+               if (blobContainer == null) {
+                   blobContainer = blobStore().blobContainer(basePath());
+                   this.blobContainer = blobContainer;
+               }
+           }
         }
+
+        return blobContainer;
     }
 
     /**
-     * Returns the BlobStore to read and write data.
+     * maintains single lazy instance of {@link BlobStore}
      */
-    protected abstract BlobStore blobStore();
+    protected BlobStore blobStore() {
+        BlobStore store = blobStore;
```

Comment:
it could be done on `generic` thread e.g. from `IndexShard.restoreFromRepository` chain:

```
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.verificationThreadCheck(BlobStoreRepository.java:608)
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.blobContainer(BlobStoreRepository.java:301)
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.listBlobsToGetLatestIndexId(BlobStoreRepository.java:809)
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.latestIndexBlobId(BlobStoreRepository.java:787)
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:650)
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:454)
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:278)
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:301)
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:276)
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1547)
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$6(IndexShard.java:2015)
```

Reply:
ok, then assert that it's either snapshot or generic threadpool

Reply:
can you address this comment, i.e. make sure that the blobStore is only accessed through the snapshot / generic thread.